internal/gps/source_cache_test.go

```diff
@@ -0,0 +1,473 @@
+			return exp[i].Ident().less(exp[j].Ident())
+		})
+		sort.Slice(got, func(i, j int) bool {
+			return got[i].Ident().less(got[j].Ident())
+		})
+		if len(exp) != len(got) {
+			t.Errorf("expected %d locked projects but got %d: expected %v\ngot %v", len(exp), len(got), exp, got)
+		}
+		for i := range exp {
+			if !exp[i].Eq(got[i]) {
+				t.Errorf("index %d: expected locked project equal to %v but got %v", i, exp[i], got[i])
+			}
+		}
+	}
+}
+
+// comparePackageTree compares two pkgtree.PackageTree and reports differences as test errors.
+func comparePackageTree(t *testing.T, exp, got pkgtree.PackageTree) {
+	if got.ImportRoot != exp.ImportRoot {
+		t.Errorf("expected package tree root %q but got %q", exp.ImportRoot, got.ImportRoot)
+	}
+	{
+		exp, got := exp.Packages, got.Packages
+		if len(exp) != len(got) {
+			t.Errorf("expected packages %v but got %v", exp, got)
+		} else {
+			for k, v := range exp {
+				if v2, ok := got[k]; !ok {
+					t.Errorf("key %q: expected %v but got none", k, v)
+				} else if !packageOrErrEqual(v, v2) {
+					t.Errorf("key %q: expected %v but got %v", k, v, v2)
+				}
+			}
+		}
+	}
+}
+
+// pairedVersionsEqual returns true if the PairedVersions have equal VersionTypes and Strings.
+func pairedVersionsEqual(a, b PairedVersion) bool {
+	return a.Type() == b.Type() && a.String() == b.String()
+}
+
+func constraintsEqual(a, b Constraint) bool {
+	if a == nil || IsAny(a) {
+		return b == nil || IsAny(b)
+	}
+	if b == nil || IsAny(b) {
+		return false
+	}
+
+	switch a.Intersect(b) {
+	case a, b:
+	default:
+		return false
+	}
+
+	switch b.Intersect(a) {
+	case a, b:
+		return true
+	default:
+		return false
+	}
+}
+
+func projectConstraintsEqual(exp, got ProjectConstraints) bool {
+	loop, check := exp, got
+	if len(got) > len(exp) {
+		loop, check = got, exp
+	}
+	for pr, pp := range loop {
+		pp2, ok := check[pr]
+		if !ok {
+			return false
+		}
+		if pp.Source != pp2.Source {
+			return false
+		}
+		if !constraintsEqual(pp.Constraint, pp2.Constraint) {
+			return false
+		}
+	}
+	return true
+}
+
+func mapStringBoolEqual(exp, got map[string]bool) bool {
+	loop, check := exp, got
+	if len(got) > len(exp) {
+		loop, check = got, exp
+	}
+	for k, v := range loop {
+		v2, ok := check[k]
+		if !ok || v != v2 {
+			return false
+		}
+	}
+	return true
+}
+
+func safeError(err error) string {
+	if err == nil {
+		return ""
+	}
+	return err.Error()
+}
+
+// packageOrErrEqual return true if the pkgtree.PackageOrErrs are equal. Error equality is
+// string based. Imports and TestImports are treated as sets, and will be sorted.
+func packageOrErrEqual(a, b pkgtree.PackageOrErr) bool {
+	if safeError(a.Err) != safeError(b.Err) {
+		return false
+	}
+	if a.P.Name != b.P.Name {
+		return false
+	}
+	if a.P.ImportPath != b.P.ImportPath {
+		return false
+	}
+	if a.P.CommentPath != b.P.CommentPath {
+		return false
+	}
+	sort.Strings(a.P.Imports)
+	sort.Strings(b.P.Imports)
+	if len(a.P.Imports) != len(b.P.Imports) {
```

Comment:
nit, but let's sort after the length comparison so that we don't do an unnecessary sort.